### PR TITLE
dmd: made derivation compile when overriding to certain older versions

### PR DESCRIPTION
**Pull request by dukc**
I ported legacy compatibility improvements from [oldDDerivations](https://github.com/dukc/oldDDerivations). These let dmd to build if the user overrides version/hashes to DMD version 2.084.1, 2.087.1 or 2.088.1, and likely some other versions of around that age.

Also added myself as DMD maintainer - if you do not want me, tell and I remove myself again.

###### Motivation for this change

@lionello [mentioned earlier](https://github.com/NixOS/nixpkgs/pull/127932#discussion_r694052350) he wants to make it easy for the user to override the versions/hashes as needed.

###### Things done

I have already tested the derivation on oldDDerivations, using it with the official channels. I also test-built this PR with DMD version 2.084.1. You can test yourself with `nix-build -E 'with import ./default.nix {}; dmd.override {version = "...";dmdSha256 = "...";druntimeSha256 = "...";phobosSha256 = "...";}'` at root directory - the quickest way to get the versions/hashes is oldDDerivations.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dukc** [comment]
ping @ThomasMader  @lionello

**Review by lionello — APPROVED** [review]

File: pkgs/development/compilers/dmd/default.nix [hunk 1]
```diff
@@ -60,24 +60,46 @@ stdenv.mkDerivation rec {
...
 
-  '' + lib.optionalString (version == "2.092.1") ''
+  + lib.optionalString (version == "2.092.1") ''
     rm dmd/test/dshell/test6952.d
   '' + lib.optionalString (builtins.compareVersions "2.092.1" version < 0) ''
     substituteInPlace dmd/test/dshell/test6952.d --replace "/usr/bin/env bash" "${bash}/bin/bash"
+  ''
 
-  '' + ''
+  + ''
```
Comment (otavio) [inline pkgs/development/compilers/dmd/default.nix:102]: Unrelated style change. Please undo.

File: pkgs/development/compilers/dmd/default.nix [hunk 2]
```diff
@@ -87,8 +109,9 @@ stdenv.mkDerivation rec {
     rm dmd/test/runnable/gdb15729.sh
     rm dmd/test/runnable/gdb4149.d
     rm dmd/test/runnable/gdb4181.d
+  ''
 
-  '' + lib.optionalString stdenv.isLinux ''
+  + lib.optionalString stdenv.isLinux ''
```
Comment (otavio) [inline pkgs/development/compilers/dmd/default.nix:114]: So does here.

**Review by otavio — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/dmd/default.nix (hunk 1 above)
Comment (dukc) [reply pkgs/development/compilers/dmd/default.nix:102]: I did that on purpose. I added a multi-line comment some way up, so I feel I need some extra space between the Bash blocks to keep them visually separate from each other. The old style is not fit for that purpose IMO.

**Comment by dukc** [comment]
ping @ThomasMader

**Comment by Artturin** [comment]
Result of `nixpkgs-review pr 152988` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dmd</li>
  </ul>
</details>

**Comment by dukc** [comment]
@Artturin thanks for bringing this up again. What's still required if we want this in?

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -3168,6 +3168,12 @@
       fingerprint = "389A 78CB CD88 5E0C 4701  DEB9 FD42 C7D0 D414 94C8";
     }];
   };
+  dukc = {
+    email = "ajieskola@gmail.com";
+    github = "dukc";
+    githubId = 24233408;
+    name = "Ate Eskola";
+  };
```
Comment (Artturin) [inline maintainers/maintainer-list.nix:3176]: split this in to a separate commit before the main commit called `maintainers: add dukc`

File: pkgs/development/compilers/dmd/default.nix [hunk 4]
```diff
@@ -60,24 +60,46 @@ stdenv.mkDerivation rec {
   # Not using patches option to make it easy to patch, for example, dmd and
   # Phobos at same time if that's required
   patchPhase =
-  lib.optionalString (builtins.compareVersions version "2.092.1" <= 0) ''
+
+  # Migrates D1-style operator overloads in DMD source, to allow building with
+  # a newer DMD
+  lib.optionalString (builtins.compareVersions version "2.088.0" < 0) ''
```
Comment (Artturin) [inline pkgs/development/compilers/dmd/default.nix:66]: there's `lib.versionAtLeast` and `lib.versionOlder` in nixpkgs so use those if possible

https://nixos.org/manual/nixpkgs/stable/#sec-functions-library-strings

Comment (dukc) [reply pkgs/development/compilers/dmd/default.nix]: Done

File: maintainers/maintainer-list.nix (hunk 3 above)
Comment (dukc) [reply maintainers/maintainer-list.nix:3176]: I split into separate commit, but it ended up after, not before, the main commit. I tried to swap their order once, screwed up somehow and had to do it again.

Comment (Artturin) [reply maintainers/maintainer-list.nix:3176]: Adding yourself as a maintainer to dmd should be in a separate commit from maintainers list one

I recommend learning git rebase

File: pkgs/development/compilers/dmd/default.nix [hunk 5]
```diff
@@ -63,15 +63,15 @@ stdenv.mkDerivation rec {
 
   # Migrates D1-style operator overloads in DMD source, to allow building with
   # a newer DMD
-  lib.optionalString (builtins.compareVersions version "2.088.0" < 0) ''
+  lib.optionalString (lib.versionOlder version "2.088.0") ''
     patch -p1 -F3 --directory=dmd -i ${(fetchpatch {
```
Comment (Artturin) [inline pkgs/development/compilers/dmd/default.nix:67]: This commit should be squashed

File: maintainers/maintainer-list.nix (hunk 3 above)
Comment (dukc) [reply maintainers/maintainer-list.nix:3176]: Ok tried again, better luck this time.

File: pkgs/development/compilers/dmd/default.nix (hunk 5 above)
Comment (dukc) [reply pkgs/development/compilers/dmd/default.nix:67]: Done
